### PR TITLE
Fix segfault in MathUnary

### LINKED ISSUE #51495: Segmentation fault: DB::FunctionMathUnary<DB::UnaryFunctionVectorized<... ErfName, &erf>>
``` sql
CREATE TABLE t10 (`c0` Int32) ENGINE = MergeTree ORDER BY tuple();
INSERT INTO t10 (c0) FORMAT Values (-1);
SELECT 1 FROM t10 GROUP BY erf(-sign(t10.c0)); -- crash
SELECT 1 FROM t10 GROUP BY -sign(t10.c0); -- logical error
```
https://s3.amazonaws.com/clickhouse-test-reports/51482/ddca9ece501e7e5554fbf439008f990997c7f2d1/sqlancer__release_.html

```
2023.06.27 15:07:21.473621 [ 470 ] {} <Fatal> BaseDaemon: ########################################
2023.06.27 15:07:21.473661 [ 470 ] {} <Fatal> BaseDaemon: (version 23.6.1.1313, build id: 3862F5999E824C3519B29E696F5C9476D4868E15, git hash: 9a30195244c7c682cb21c467fcb3580abd34b382) (from thread 353) (query_id: 95126de5-ebbf-42fd-85a8-bb9a9c4f3606) (query: SELECT (- ((erf ((- (-40537385)))))) FROM t3 GROUP BY (erf ((- ((sign (t3.c0))))))) Received signal Segmentation fault (11)
2023.06.27 15:07:21.473678 [ 470 ] {} <Fatal> BaseDaemon: Address: 0x10. Access: read. Address not mapped to object.
2023.06.27 15:07:21.473693 [ 470 ] {} <Fatal> BaseDaemon: Stack trace: 0x000000000a342750 0x00000000087f458a 0x00000000087f3b2e 0x00000000126238ef 0x0000000012624362 0x0000000012625659 0x0000000012da82ac 0x0000000014dcc944 0x0000000014f209ce 0x0000000014fce246 0x0000000014fddc42 0x0000000014fbc637 0x0000000014f4bcbc 0x0000000014f4abce 0x00000000139f5fae 0x0000000013d23753 0x0000000013d2848c 0x0000000014af8055 0x0000000014afc304 0x0000000014b6c4b2 0x0000000017ad6b54 0x0000000017ad7d71 0x0000000017c54b27 0x0000000017c5255c 0x00007fb7ca42ab43 0x00007fb7ca4bca00
2023.06.27 15:07:22.586605 [ 470 ] {} <Fatal> BaseDaemon: 2. DB::FunctionMathUnary<DB::UnaryFunctionVectorized<DB::(anonymous namespace)::ErfName, &erf>>::executeImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x000000000a342750 in /clickhouse
2023.06.27 15:07:23.538328 [ 470 ] {} <Fatal> BaseDaemon: 3. DB::IFunction::executeImplDryRun(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x00000000087f458a in /clickhouse
2023.06.27 15:07:24.442075 [ 470 ] {} <Fatal> BaseDaemon: 4. DB::FunctionToExecutableFunctionAdaptor::executeDryRunImpl(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0x00000000087f3b2e in /clickhouse
2023.06.27 15:07:24.451789 [ 470 ] {} <Fatal> BaseDaemon: 5. ./build_docker/./src/Functions/IFunction.cpp:0: DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x00000000126238ef in /clickhouse
2023.06.27 15:07:24.460625 [ 470 ] {} <Fatal> BaseDaemon: 6.1. inlined from ./build_docker/./contrib/boost/boost/smart_ptr/intrusive_ptr.hpp:115: intrusive_ptr
2023.06.27 15:07:24.460660 [ 470 ] {} <Fatal> BaseDaemon: 6.2. inlined from ./build_docker/./contrib/boost/boost/smart_ptr/intrusive_ptr.hpp:122: boost::intrusive_ptr<DB::IColumn const>::operator=(boost::intrusive_ptr<DB::IColumn const>&&)
2023.06.27 15:07:24.460672 [ 470 ] {} <Fatal> BaseDaemon: 6.3. inlined from ./build_docker/./src/Common/COW.h:136: COW<DB::IColumn>::immutable_ptr<DB::IColumn>::operator=(COW<DB::IColumn>::immutable_ptr<DB::IColumn>&&)
2023.06.27 15:07:24.460682 [ 470 ] {} <Fatal> BaseDaemon: 6. ./build_docker/./src/Functions/IFunction.cpp:302: DB::IExecutableFunction::executeWithoutSparseColumns(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012624362 in /clickhouse
2023.06.27 15:07:24.470130 [ 470 ] {} <Fatal> BaseDaemon: 7.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/vector:434: ~vector
2023.06.27 15:07:24.470163 [ 470 ] {} <Fatal> BaseDaemon: 7. ./build_docker/./src/Functions/IFunction.cpp:375: DB::IExecutableFunction::execute(std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>> const&, std::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x0000000012625659 in /clickhouse
2023.06.27 15:07:24.539474 [ 470 ] {} <Fatal> BaseDaemon: 8.1. inlined from ./build_docker/./src/Interpreters/ActionsDAG.cpp:0: DB::executeActionForHeader(DB::ActionsDAG::Node const*, std::vector<DB::ColumnWithTypeAndName, std::allocator<DB::ColumnWithTypeAndName>>)
2023.06.27 15:07:24.539513 [ 470 ] {} <Fatal> BaseDaemon: 8. ./build_docker/./src/Interpreters/ActionsDAG.cpp:656: DB::ActionsDAG::updateHeader(DB::Block) const @ 0x0000000012da82ac in /clickhouse
2023.06.27 15:07:24.544307 [ 470 ] {} <Fatal> BaseDaemon: 9. ./build_docker/./src/Processors/Transforms/ExpressionTransform.cpp:0: DB::ExpressionTransform::transformHeader(DB::Block, DB::ActionsDAG const&) @ 0x0000000014dcc944 in /clickhouse
2023.06.27 15:07:24.554175 [ 470 ] {} <Fatal> BaseDaemon: 10. ./build_docker/./src/Processors/QueryPlan/ExpressionStep.cpp:32: DB::ExpressionStep::ExpressionStep(DB::DataStream const&, std::shared_ptr<DB::ActionsDAG> const&) @ 0x0000000014f209ce in /clickhouse
2023.06.27 15:07:24.569595 [ 470 ] {} <Fatal> BaseDaemon: 11. ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:714: std::__unique_if<DB::ExpressionStep>::__unique_single std::make_unique[abi:v15000]<DB::ExpressionStep, DB::DataStream const&, std::shared_ptr<DB::ActionsDAG>&>(DB::DataStream const&, std::shared_ptr<DB::ActionsDAG>&) @ 0x0000000014fce246 in /clickhouse
2023.06.27 15:07:24.603756 [ 470 ] {} <Fatal> BaseDaemon: 12. ./build_docker/./src/Processors/QueryPlan/Optimizations/optimizeUseAggregateProjection.cpp:0: DB::QueryPlanOptimizations::optimizeUseAggregateProjections(DB::QueryPlan::Node&, std::list<DB::QueryPlan::Node, std::allocator<DB::QueryPlan::Node>>&) @ 0x0000000014fddc42 in /clickhouse
2023.06.27 15:07:24.609514 [ 470 ] {} <Fatal> BaseDaemon: 13. ./build_docker/./src/Processors/QueryPlan/Optimizations/optimizeTree.cpp:129: DB::QueryPlanOptimizations::optimizeTreeSecondPass(DB::QueryPlanOptimizationSettings const&, DB::QueryPlan::Node&, std::list<DB::QueryPlan::Node, std::allocator<DB::QueryPlan::Node>>&) @ 0x0000000014fbc637 in /clickhouse
2023.06.27 15:07:24.631508 [ 470 ] {} <Fatal> BaseDaemon: 14. ./build_docker/./src/Processors/QueryPlan/QueryPlan.cpp:486: DB::QueryPlan::optimize(DB::QueryPlanOptimizationSettings const&) @ 0x0000000014f4bcbc in /clickhouse
2023.06.27 15:07:24.649052 [ 470 ] {} <Fatal> BaseDaemon: 15.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/compressed_pair.h:36: __compressed_pair_elem
2023.06.27 15:07:24.649079 [ 470 ] {} <Fatal> BaseDaemon: 15.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/compressed_pair.h:108: __compressed_pair<std::__value_init_tag, std::__value_init_tag>
2023.06.27 15:07:24.649095 [ 470 ] {} <Fatal> BaseDaemon: 15.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:168: unique_ptr<true, void>
2023.06.27 15:07:24.649105 [ 470 ] {} <Fatal> BaseDaemon: 15. ./build_docker/./src/Processors/QueryPlan/QueryPlan.cpp:169: DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&) @ 0x0000000014f4abce in /clickhouse
2023.06.27 15:07:24.671201 [ 470 ] {} <Fatal> BaseDaemon: 16. ./build_docker/./src/Interpreters/InterpreterSelectWithUnionQuery.cpp:379: DB::InterpreterSelectWithUnionQuery::execute() @ 0x00000000139f5fae in /clickhouse
2023.06.27 15:07:24.696568 [ 470 ] {} <Fatal> BaseDaemon: 17. ./build_docker/./src/Interpreters/executeQuery.cpp:0: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x0000000013d23753 in /clickhouse
2023.06.27 15:07:24.724485 [ 470 ] {} <Fatal> BaseDaemon: 18.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:612: shared_ptr
2023.06.27 15:07:24.724516 [ 470 ] {} <Fatal> BaseDaemon: 18.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:723: std::shared_ptr<DB::IAST>::operator=[abi:v15000](std::shared_ptr<DB::IAST>&&)
2023.06.27 15:07:24.724537 [ 470 ] {} <Fatal> BaseDaemon: 18.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/tuple:533: void std::__memberwise_forward_assign[abi:v15000]<std::tuple<std::shared_ptr<DB::IAST>&, DB::BlockIO&>, std::tuple<std::shared_ptr<DB::IAST>, DB::BlockIO>, std::shared_ptr<DB::IAST>, DB::BlockIO, 0ul, 1ul>(std::tuple<std::shared_ptr<DB::IAST>&, DB::BlockIO&>&, std::tuple<std::shared_ptr<DB::IAST>, DB::BlockIO>&&, std::__tuple_types<std::shared_ptr<DB::IAST>, DB::BlockIO>, std::__tuple_indices<0ul, 1ul>)
2023.06.27 15:07:24.724563 [ 470 ] {} <Fatal> BaseDaemon: 18.4. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/tuple:1138: std::tuple<std::shared_ptr<DB::IAST>&, DB::BlockIO&>& std::tuple<std::shared_ptr<DB::IAST>&, DB::BlockIO&>::operator=[abi:v15000]<std::shared_ptr<DB::IAST>, DB::BlockIO, 0>(std::tuple<std::shared_ptr<DB::IAST>, DB::BlockIO>&&)
2023.06.27 15:07:24.724573 [ 470 ] {} <Fatal> BaseDaemon: 18. ./build_docker/./src/Interpreters/executeQuery.cpp:1262: DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::shared_ptr<DB::Context>, std::function<void (DB::QueryResultDetails const&)>, std::optional<DB::FormatSettings> const&) @ 0x0000000013d2848c in /clickhouse
2023.06.27 15:07:24.747687 [ 470 ] {} <Fatal> BaseDaemon: 19.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/optional:260: ~__optional_destruct_base
2023.06.27 15:07:24.747724 [ 470 ] {} <Fatal> BaseDaemon: 19. ./build_docker/./src/Server/HTTPHandler.cpp:835: DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::optional<DB::CurrentThread::QueryScope>&) @ 0x0000000014af8055 in /clickhouse
2023.06.27 15:07:24.775151 [ 470 ] {} <Fatal> BaseDaemon: 20. ./build_docker/./src/Server/HTTPHandler.cpp:0: DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&) @ 0x0000000014afc304 in /clickhouse
2023.06.27 15:07:24.777669 [ 470 ] {} <Fatal> BaseDaemon: 21.1. inlined from ./build_docker/./base/poco/Foundation/include/Poco/AutoPtr.h:205: Poco::AutoPtr<Poco::Net::HTTPServerParams>::operator->()
2023.06.27 15:07:24.777701 [ 470 ] {} <Fatal> BaseDaemon: 21. ./build_docker/./src/Server/HTTP/HTTPServerConnection.cpp:69: DB::HTTPServerConnection::run() @ 0x0000000014b6c4b2 in /clickhouse
2023.06.27 15:07:24.779251 [ 470 ] {} <Fatal> BaseDaemon: 22. ./build_docker/./base/poco/Net/src/TCPServerConnection.cpp:57: Poco::Net::TCPServerConnection::start() @ 0x0000000017ad6b54 in /clickhouse
2023.06.27 15:07:24.781520 [ 470 ] {} <Fatal> BaseDaemon: 23.1. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:48: std::default_delete<Poco::Net::TCPServerConnection>::operator()[abi:v15000](Poco::Net::TCPServerConnection*) const
2023.06.27 15:07:24.781548 [ 470 ] {} <Fatal> BaseDaemon: 23.2. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:305: std::unique_ptr<Poco::Net::TCPServerConnection, std::default_delete<Poco::Net::TCPServerConnection>>::reset[abi:v15000](Poco::Net::TCPServerConnection*)
2023.06.27 15:07:24.781561 [ 470 ] {} <Fatal> BaseDaemon: 23.3. inlined from ./build_docker/./contrib/llvm-project/libcxx/include/__memory/unique_ptr.h:259: ~unique_ptr
2023.06.27 15:07:24.781571 [ 470 ] {} <Fatal> BaseDaemon: 23. ./build_docker/./base/poco/Net/src/TCPServerDispatcher.cpp:116: Poco::Net::TCPServerDispatcher::run() @ 0x0000000017ad7d71 in /clickhouse
2023.06.27 15:07:24.784176 [ 470 ] {} <Fatal> BaseDaemon: 24. ./build_docker/./base/poco/Foundation/src/ThreadPool.cpp:202: Poco::PooledThread::run() @ 0x0000000017c54b27 in /clickhouse
2023.06.27 15:07:24.786437 [ 470 ] {} <Fatal> BaseDaemon: 25.1. inlined from ./build_docker/./base/poco/Foundation/include/Poco/SharedPtr.h:231: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::get()
2023.06.27 15:07:24.786460 [ 470 ] {} <Fatal> BaseDaemon: 25.2. inlined from ./build_docker/./base/poco/Foundation/include/Poco/SharedPtr.h:139: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::assign(Poco::Runnable*)
2023.06.27 15:07:24.786481 [ 470 ] {} <Fatal> BaseDaemon: 25.3. inlined from ./build_docker/./base/poco/Foundation/include/Poco/SharedPtr.h:180: Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable>>::operator=(Poco::Runnable*)
2023.06.27 15:07:24.786490 [ 470 ] {} <Fatal> BaseDaemon: 25. ./build_docker/./base/poco/Foundation/src/Thread_POSIX.cpp:350: Poco::ThreadImpl::runnableEntry(void*) @ 0x0000000017c5255c in /clickhouse
2023.06.27 15:07:24.786506 [ 470 ] {} <Fatal> BaseDaemon: 26. ? @ 0x00007fb7ca42ab43 in ?
2023.06.27 15:07:24.786516 [ 470 ] {} <Fatal> BaseDaemon: 27. ? @ 0x00007fb7ca4bca00 in ?
2023.06.27 15:07:24.786527 [ 470 ] {} <Fatal> BaseDaemon: Integrity check of the executable skipped because the reference checksum could not be read.
2023.06.27 15:07:24.786547 [ 470 ] {} <Fatal> BaseDaemon: This ClickHouse version is not official and should be upgraded to the official build.
2023.06.27 15:07:24.786635 [ 470 ] {} <Fatal> BaseDaemon: No settings were changed
2023.06.27 15:07:25.798195 [ 29 ] {} <Fatal> Application: Child process was terminated by signal 11.
```

```
(lldb) bt
* thread #3, name = 'TCPHandler', stop reason = signal SIGSEGV: invalid address (fault address: 0x18)
  * frame #0: 0x000000001ad78d6c clickhouse`DB::PODArrayBase<1ul, 4096ul, Allocator<false, false>, 63ul, 64ul>::size(this=0x0000000000000010) const at PODArray.h:222:35
    frame #1: 0x000000001df11b91 clickhouse`COW<DB::IColumn>::immutable_ptr<DB::IColumn> DB::FunctionMathUnary<DB::UnaryFunctionVectorized<DB::(anonymous namespace)::ErfName, &erf>>::execute<signed char, double>(col=0x0000000000000000) at FunctionMathUnary.h:112:38
    frame #2: 0x000000001df10779 clickhouse`bool DB::FunctionMathUnary<DB::UnaryFunctionVectorized<DB::(anonymous namespace)::ErfName, &erf>>::executeImpl(this=0x00007ffecf5e9908, types=0x00007ffecf5e97f8) const::'lambda'(auto const&)::operator()<DB::TypePair<void, signed char>>(auto const&) const at FunctionMathUnary.h:157:27
    frame #3: 0x000000001df10257 clickhouse`bool DB::callOnBasicType<void, true, true, true, false, DB::FunctionMathUnary<DB::UnaryFunctionVectorized<DB::(anonymous namespace)::ErfName, &erf>>::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName>> const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const::'lambda'(auto const&)&>(number=Int8, f=0x00007ffecf5e9908) at callOnTypeIndex.h:32:50
    frame #4: 0x000000001df0fd82 clickhouse`DB::FunctionMathUnary<DB::UnaryFunctionVectorized<DB::(anonymous namespace)::ErfName, &erf>>::executeImpl(this=0x00007ffe4f8cf778, arguments=size=1, (null)=std::__1::shared_ptr<const DB::IDataType>::element_type @ 0x00007ffec69c95b8 strong=9 weak=2, (null)=0) const at FunctionMathUnary.h:160:14
    frame #5: 0x000000001ad75279 clickhouse`DB::IFunction::executeImplDryRun(this=0x00007ffe4f8cf778, arguments=size=1, result_type=std::__1::shared_ptr<const DB::IDataType>::element_type @ 0x00007ffec69c95b8 strong=9 weak=2, input_rows_count=0) const at IFunction.h:430:16
    frame #6: 0x000000001ad7336d clickhouse`DB::FunctionToExecutableFunctionAdaptor::executeDryRunImpl(this=0x00007ffe4f8e57c0, arguments=size=1, result_type=std::__1::shared_ptr<const DB::IDataType>::element_type @ 0x00007ffec69c95b8 strong=9 weak=2, input_rows_count=0) const at IFunctionAdaptors.h:26:26
    frame #7: 0x000000002b458e06 clickhouse`DB::IExecutableFunction::executeWithoutLowCardinalityColumns(this=0x00007ffe4f8e57c0, args=size=1, result_type=std::__1::shared_ptr<const DB::IDataType>::element_type @ 0x00007ffec69c95b8 strong=9 weak=2, input_rows_count=0, dry_run=true) const at IFunction.cpp:245:15
    frame #8: 0x000000002b459ee6 clickhouse`DB::IExecutableFunction::executeWithoutSparseColumns(this=0x00007ffe4f8e57c0, arguments=size=1, result_type=std::__1::shared_ptr<const DB::IDataType>::element_type @ 0x00007ffec69c95b8 strong=9 weak=2, input_rows_count=0, dry_run=true) const at IFunction.cpp:302:22
    frame #9: 0x000000002b45add5 clickhouse`DB::IExecutableFunction::execute(this=0x00007ffe4f8e57c0, arguments=size=1, result_type=std::__1::shared_ptr<const DB::IDataType>::element_type @ 0x00007ffec69c95b8 strong=9 weak=2, input_rows_count=0, dry_run=true) const at IFunction.cpp:374:16
    frame #10: 0x000000002c0cffa8 clickhouse`DB::executeActionForHeader(node=0x00007ffe45b1d7b0, arguments=size=1) at ActionsDAG.cpp:537:49
    frame #11: 0x000000002c0cf9a8 clickhouse`DB::ActionsDAG::updateHeader(this=0x00007ffec69c9910, header=Block @ 0x00007ffecf5ea468) const at ActionsDAG.cpp:656:44
    frame #12: 0x000000002f1245c7 clickhouse`DB::ExpressionTransform::transformHeader(header=<unavailable>, expression=0x00007ffec69c9910) at ExpressionTransform.cpp:8:23
    frame #13: 0x000000002f3838a6 clickhouse`DB::ExpressionStep::ExpressionStep(this=0x00007ffe4f977a40, input_stream_=0x00007ffe33735da0, actions_dag_=std::__1::shared_ptr<DB::ActionsDAG>::element_type @ 0x00007ffec69c9910 strong=1 weak=1) at ExpressionStep.cpp:31:9
    frame #14: 0x000000002b47d811 clickhouse`std::__1::__unique_if<DB::ExpressionStep>::__unique_single std::__1::make_unique[abi:v15000]<DB::ExpressionStep, DB::DataStream const&, std::__1::shared_ptr<DB::ActionsDAG>&>(__args=0x00007ffe33735da0, __args=std::__1::shared_ptr<DB::ActionsDAG>::element_type @ 0x00007ffec69c9910 strong=1 weak=1) at unique_ptr.h:714:32
    frame #15: 0x000000002f4b9da6 clickhouse`DB::QueryPlanOptimizations::optimizeUseAggregateProjections(node=0x00007ffe45981470, nodes=size=8) at optimizeUseAggregateProjection.cpp:684:36
    frame #16: 0x000000002f47a776 clickhouse`DB::QueryPlanOptimizations::optimizeTreeSecondPass(optimization_settings=0x00007ffecf5eb710, root=0x00007ffe45985d00, nodes=size=8) at optimizeTree.cpp:129:47
    frame #17: 0x000000002f3cae3d clickhouse`DB::QueryPlan::optimize(this=0x00007ffe4f9075b8, optimization_settings=0x00007ffecf5eb710) at QueryPlan.cpp:484:5
    frame #18: 0x000000002f3ca957 clickhouse`DB::QueryPlan::buildQueryPipeline(this=0x00007ffe4f9075b8, optimization_settings=0x00007ffecf5eb710, build_pipeline_settings=0x00007ffecf5eb6a8) at QueryPlan.cpp:161:5
    frame #19: 0x000000002d21f88d clickhouse`DB::InterpreterSelectQueryAnalyzer::buildQueryPipeline(this=0x00007ffe4f907500) at InterpreterSelectQueryAnalyzer.cpp:252:34
    frame #20: 0x000000002d21f5db clickhouse`DB::InterpreterSelectQueryAnalyzer::execute(this=0x00007ffe4f907500) at InterpreterSelectQueryAnalyzer.cpp:221:29
    frame #21: 0x000000002d740737 clickhouse`DB::executeQueryImpl(begin="SELECT (- ((erf ((- (-40537385)))))) FROM t10 GROUP BY (erf ((- ((sign (t10.c0))))));", end="", context=std::__1::shared_ptr<DB::Context>::element_type @ 0x00007ffe336de000 strong=3 weak=9, internal=false, stage=Complete, istr=0x0000000000000000) at executeQuery.cpp:723:36
    frame #22: 0x000000002d73bdc4 clickhouse`DB::executeQuery(query="SELECT (- ((erf ((- (-40537385)))))) FROM t10 GROUP BY (erf ((- ((sign (t10.c0))))));", context=std::__1::shared_ptr<DB::Context>::element_type @ 0x00007ffe336de000 strong=3 weak=9, internal=false, stage=Complete) at executeQuery.cpp:1182:30
    frame #23: 0x000000002ec0278d clickhouse`DB::TCPHandler::runImpl(this=0x00007ffe45b69c00) at TCPHandler.cpp:420:24
    frame #24: 0x000000002ec143a5 clickhouse`DB::TCPHandler::run(this=0x00007ffe45b69c00) at TCPHandler.cpp:2043:9
    frame #25: 0x0000000034113ad9 clickhouse`Poco::Net::TCPServerConnection::start(this=0x00007ffe45b69c00) at TCPServerConnection.cpp:43:3
    frame #26: 0x000000003411431c clickhouse`Poco::Net::TCPServerDispatcher::run(this=0x00007ffecbb4ea00) at TCPServerDispatcher.cpp:115:20
    frame #27: 0x0000000034355634 clickhouse`Poco::PooledThread::run(this=0x00007ffff7363980) at ThreadPool.cpp:188:14
    frame #28: 0x00000000343523da clickhouse`Poco::(anonymous namespace)::RunnableHolder::run(this=0x00007ffff7809910) at Thread.cpp:45:11
    frame #29: 0x00000000343510de clickhouse`Poco::ThreadImpl::runnableEntry(pThread=0x00007ffff73639b8) at Thread_POSIX.cpp:335:27
    frame #30: 0x00007ffff7f91609 libpthread.so.0`start_thread(arg=<unavailable>) at pthread_create.c:477:8
    frame #31: 0x00007ffff7eb6163 libc.so.6`__clone + 67
(lldb) fr v
(const DB::PODArrayBase<1, 4096, Allocator<false, false>, 63, 64> *) this = 0x0000000000000010
(lldb) up
frame #1: 0x000000001df11b91 clickhouse`COW<DB::IColumn>::immutable_ptr<DB::IColumn> DB::FunctionMathUnary<DB::UnaryFunctionVectorized<DB::(anonymous namespace)::ErfName, &erf>>::execute<signed char, double>(col=0x0000000000000000) at FunctionMathUnary.h:112:38
(lldb) fr v
(const DB::ColumnVector<signed char> *) col = nullptr
(const DB::ColumnVector<signed char>::Container &) src_data = 0x0000000000000010: {
  DB::PODArrayBase<1, 4096, Allocator<false, false>, 63, 64> = (c_start = <read memory from 0x10 failed (0 of 8 bytes read)>, c_end = <read memory from 0x18 failed (0 of 8 bytes read)>, c_end_of_storage = <read memory from 0x20 failed (0 of 8 bytes read)>, mprotected = <read memory from 0x28 failed (0 of 1 bytes read)>)
}
(const size_t) size = 140732377503520
(COWHelper<DB::ColumnVectorHelper, DB::ColumnVector<double> >::MutablePtr) dst = {
  boost::intrusive_ptr<DB::ColumnVector<double> > = {
    px = nullptr
  }
}
(DB::ColumnVector<double>::Container &) dst_data = 0x00007ffe45985ea0: {
  DB::PODArrayBase<8, 4096, Allocator<false, false>, 63, 64> = (c_start = "@X\xfa\U0000001a", c_end = "", c_end_of_storage = "", mprotected = true)
}
(lldb)
```

`checkAndGetColumn<ColVecType>(col.column.get())` returned nullptr
https://github.com/ClickHouse/ClickHouse/blob/f85b88b5b8fc9dae5b4f4bdcc60102634726366c/src/Functions/FunctionMathUnary.h#L156-L157

### PR DESCRIPTION
**Pull request by qoega**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Check for incorrect column conversion. Closes https://github.com/ClickHouse/ClickHouse/issues/51495.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
This is an automated comment for commit d3b8b454f853c63da4b94ec97afdcb1528ffdc22 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/51499/d3b8b454f853c63da4b94ec97afdcb1528ffdc22/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Functions/FunctionMathUnary.h [hunk 1]
```diff
@@ -154,6 +154,8 @@ class FunctionMathUnary : public IFunction
             using ColVecType = ColumnVectorOrDecimal<Type>;
 
             const auto col_vec = checkAndGetColumn<ColVecType>(col.column.get());
+            if (col_vec == nullptr)
```
Comment (rschu1ze) [inline src/Functions/FunctionMathUnary.h:157]: Will prevent the crash but the true problem is further up the stack, I guess.

\+ needs a test

**Review by rschu1ze — APPROVED** [review]

File: src/Functions/FunctionMathUnary.h (hunk 1 above)
Comment (qoega) [reply src/Functions/FunctionMathUnary.h:157]: Currently building binary to canonize the test. But I agree that we need to find root cause.

**Review by alexey-milovidov — APPROVED** [review]

**Comment by qoega** [comment]
It looks we do not fail/retry when we could not connect 
```
+ clickhouse-client --query 'SELECT '\''Connected to clickhouse-server after attaching gdb'\'''
Code: 210. DB::NetException: Connection refused (localhost:9000). (NETWORK_ERROR)


real	0m0.018s
user	0m0.011s
sys	0m0.007s
```

**Comment by qoega** [comment]
- Stress test (tsan) https://github.com/ClickHouse/ClickHouse/pull/51563
- Stress test (ubsan) https://github.com/ClickHouse/ClickHouse/issues/51592

**Comment by qoega** [comment]
Screenshot 2023-06-29 at 18 22 55

**Pull request merged**